sql/logictest: TestSqlLiteLogic failed due to misbehaved zigzag joiner

[(sql/logictest).TestSqlLiteLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2366002&tab=buildLog) on [master@80e7127197f76ef35c1f6ec3984c4d49d4afde7f](https://github.com/cockroachdb/cockroach/commits/80e7127197f76ef35c1f6ec3984c4d49d4afde7f):

Fatal error:
```
panic: MoveToDraining called in state StateTrailingMeta with err: fetcher-mem: memory budget exceeded: 10240 bytes requested, 0 currently allocated, 0 bytes in budget
(1) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).increaseBudget
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:684
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).reserveBytes
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:617
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Grow
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:572
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Resize
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:552
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).ResizeTo
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:565
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).fetch
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:348
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).nextBatch
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:457
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*KVFetcher).NextKV
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_fetcher.go:115
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).NextKey
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:671
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).StartScanFrom
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:661
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).StartScan
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:553
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).maybeFetchInitialRow
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:905
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:930
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:163
  | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*OrderedSynchronizer).DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/ordered_synchronizer.eg.go:360
  | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*drainHelper).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:102
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:265
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:775
  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:392
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:996
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1002
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:873
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:639
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:114
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1474
Wraps: (2) fetcher-mem
Wraps: (3) candidate pg code: 53200
Wraps: (4) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/util/mon.memoryResource.NewBudgetExceededError
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/resource.go:37
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).increaseBudget
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:684
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).reserveBytes
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:617
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Grow
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:572
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Resize
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:552
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).ResizeTo
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:565
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).fetch
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:348
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).nextBatch
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:457
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*KVFetcher).NextKV
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_fetcher.go:115
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).NextKey
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:671
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).StartScanFrom
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:661
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).StartScan
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:553
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).maybeFetchInitialRow
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:905
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:930
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:163
  | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*OrderedSynchronizer).DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/ordered_synchronizer.eg.go:360
  | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*drainHelper).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:102
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:265
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:775
  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:392
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:996
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1002
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:873
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:639
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:114
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1474
Wraps: (5) memory budget exceeded: 10240 bytes requested, 0 currently allocated, 0 bytes in budget
Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pgerror.withCandidateCode (4) *withstack.withStack (5) *errutil.leafError [recovered]
	panic: MoveToDraining called in state StateTrailingMeta with err: fetcher-mem: memory budget exceeded: 10240 bytes requested, 0 currently allocated, 0 bytes in budget
(1) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).increaseBudget
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:684
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).reserveBytes
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:617
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Grow
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:572
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Resize
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:552
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).ResizeTo
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:565
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).fetch
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:348
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).nextBatch
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:457
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*KVFetcher).NextKV
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_fetcher.go:115
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).NextKey
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:671
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).StartScanFrom
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:661
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).StartScan
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:553
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).maybeFetchInitialRow
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:905
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:930
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:163
  | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*OrderedSynchronizer).DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/ordered_synchronizer.eg.go:360
  | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*drainHelper).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:102
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:265
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:775
  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:392
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:996
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1002
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:873
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:639
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:114
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1474
Wraps: (2) fetcher-mem
Wraps: (3) candidate pg code: 53200
Wraps: (4) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/util/mon.memoryResource.NewBudgetExceededError
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/resource.go:37
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).increaseBudget
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:684
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).reserveBytes
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:617
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Grow
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:572
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Resize
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:552
  | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).ResizeTo
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:565
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).fetch
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:348
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).nextBatch
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:457
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*KVFetcher).NextKV
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_fetcher.go:115
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).NextKey
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:671
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).StartScanFrom
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:661
  | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).StartScan
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:553
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).maybeFetchInitialRow
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:905
  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:930
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:163
  | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*OrderedSynchronizer).DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/ordered_synchronizer.eg.go:360
  | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*drainHelper).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:102
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653
  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:265
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170
  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:775
  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:392
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:996
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1002
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:873
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:639
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:114
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1474
Wraps: (5) memory budget exceeded: 10240 bytes requested, 0 currently allocated, 0 bytes in budget
Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pgerror.withCandidateCode (4) *withstack.withStack (5) *errutil.leafError

```

Stack:
```
goroutine 12405 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc007804000, 0x5816660, 0xc00621ddc0, 0x44e18e0, 0xc018da0d80)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:824 +0x359
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc007804000, 0x5816660, 0xc00621ddc0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:507 +0x61
panic(0x44e18e0, 0xc018da0d80)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/util/log.ReportOrPanic(0x5816660, 0xc00e4f7480, 0xc000c40000, 0x48c8e2d, 0x2f, 0xc00be55208, 0x2, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:340 +0x212
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).MoveToDraining(0xc014351400, 0x5787e80, 0xc018da0b80)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:601 +0x307
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).Next(0xc014351400, 0x0, 0x0, 0xc00be553b0, 0x6824a6)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:931 +0x170
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper(0xc008873800, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653 +0xa0
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).DrainMeta(0xc008873800, 0x5816660, 0xc00e4f7480, 0x1, 0xc00be554b0, 0x6bc45e)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:163 +0x5b
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta(0xc005bb9700, 0x4, 0x4, 0x5816660, 0xc00e4f7480, 0x2, 0x2, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340 +0xba
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*OrderedSynchronizer).DrainMeta(0xc0017f6a00, 0x5816660, 0xc00e4f7480, 0x5971aa0, 0xc00d914fc0, 0xc00be55620)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/ordered_synchronizer.eg.go:360 +0x151
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta(0xc0178702a0, 0x1, 0x1, 0x5816660, 0xc00e4f7480, 0x0, 0x1, 0xc0178702f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340 +0xba
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*drainHelper).Next(0xc012ba0dc0, 0x0, 0xc00be55700, 0xc00be55740, 0x24f2675)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:102 +0x19b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper(0xc001bf0b00, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653 +0xa0
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next(0xc001bf0b00, 0x5816660, 0xc00e4f7480, 0x31181b0, 0x581af20)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:265 +0xe4
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run(0x5816660, 0xc00e4f7480, 0x583f920, 0xc001bf0b00, 0x57ee3e0, 0xc0086a5880)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170 +0x35
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run(0xc001bf0b00, 0x5816660, 0xc00e4f7480)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:775 +0x92
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc005572900, 0x5816660, 0xc00e4f7480, 0x4de62a0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:392 +0x24d
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc003754d10, 0xc016971ce0, 0xc0102610e0, 0xc002898800, 0xc0086a5880, 0xc007804518, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422 +0x5d9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc003754d10, 0x5816720, 0xc01e6fde30, 0xc007804518, 0xc016971ce0, 0xc0102610e0, 0x5818de0, 0xc010b85d60, 0x0, 0xc0086a5880, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:996 +0x1d1
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc007804000, 0x5816720, 0xc01e6fde30, 0xc007804418, 0x3, 0x14ddcb474ef8, 0xc002245880, 0xc00be56100, 0xc008923eb8, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1002 +0x3c0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc007804000, 0x5816720, 0xc01e6fde30, 0xc007804418, 0x14ddcb474ef8, 0xc002245880, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:873 +0x6ff
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc007804000, 0x5816720, 0xc01e6fde30, 0x5835260, 0xc001c70500, 0xc008df8724, 0xa6, 0x0, 0x2, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:639 +0xbb1
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc007804000, 0x5816720, 0xc01e6fde30, 0x5835260, 0xc001c70500, 0xc008df8724, 0xa6, 0x0, 0x2, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:114 +0x924
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0x5835260, 0xc001c70500, 0xc008df8724, 0xa6, 0x0, 0x2, 0x21191eb3, 0xed719d4c1, 0x0, 0x21192542, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1474 +0x2d5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc007804000, 0x5816720, 0xc01e6fddd0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1476 +0x148d
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc007804000, 0x5816660, 0xc00621ddc0, 0xc00300bea0, 0x5400, 0x15000, 0xc00300bf40, 0xc006b92740, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1400 +0x1ce
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc001cdb700, 0x5816660, 0xc00621ddc0, 0xc007804000, 0x5400, 0x15000, 0xc00300bf40, 0xc006b92740, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:509 +0x104
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc0049bf5c2, 0xc001957120, 0x5816660, 0xc00621ddc0, 0xc006b92740, 0xc001cdb700, 0xc002245200, 0x58431a0, 0xc006474db0, 0xc00870c840, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:627 +0x317
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:555 +0x17e

```

<details><summary>Log preceding fatal error</summary><p>

```
*   |   | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).NextKey
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:671
*   |   | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).StartScanFrom
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:661
*   |   | github.com/cockroachdb/cockroach/pkg/sql/row.(*Fetcher).StartScan
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/fetcher.go:553
*   |   | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).maybeFetchInitialRow
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:905
*   |   | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*zigzagJoiner).Next
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/zigzagjoiner.go:930
*   |   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653
*   |   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).DrainMeta
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:163
*   |   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340
*   |   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*OrderedSynchronizer).DrainMeta
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/ordered_synchronizer.eg.go:360
*   |   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340
*   |   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*drainHelper).Next
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:102
*   |   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:653
*   |   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:265
*   |   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170
*   |   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:775
*   |   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:392
*   |   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422
*   |   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:996
*   |   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1002
*   |   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:873
*   |   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:639
*   |   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:114
*   |   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   |   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1474
*   | Wraps: (5) memory budget exceeded: 10240 bytes requested, 0 currently allocated, 0 bytes in budget
*   | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pgerror.withCandidateCode (4) *withstack.withStack (5) *errutil.leafError
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) *withstack.withStack (5) *secondary.withSecondaryError (6) *errutil.leafError
*

```

</p></details>

<details><summary>More</summary><p>


```
make stressrace TESTS=TestSqlLiteLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSqlLiteLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/rowexec/zigzagjoiner.go
+++ b/pkg/sql/rowexec/zigzagjoiner.go
@@ -927,11 +927,11 @@ func (z *zigzagJoiner) maybeFetchInitialRow() error {
 
 // Next is part of the RowSource interface.
 func (z *zigzagJoiner) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetadata) {
-	if err := z.maybeFetchInitialRow(); err != nil {
-		z.MoveToDraining(err)
-	}
-
 	for z.State == execinfra.StateRunning {
+		if err := z.maybeFetchInitialRow(); err != nil {
+			z.MoveToDraining(err)
+			break
+		}
 		row, err := z.nextRow(z.Ctx, z.FlowCtx.Txn)
 		if err != nil {
 			z.MoveToDraining(err)

